### DISCUSSION TITLE
Getting backref relationship when only setting foreign key

### DISCUSSION THREAD

**Discussion by anushree-agrawal** [sqlalchemy/sqlalchemy#8055] [Usage Questions] [answered]
## Describe the issue
When creating a new object (child) with a foreign key reference and a relationship to another object (parent), I didn't set the relationship (parent=parent) and only set the foreign key (parent_id=parent.id). 

When I tried to access `parent.child` after a `db.session.flush()`, I get an empty array instead of the newly created child object.

## To reproduce
Model classes:
```
class Parent(db.Model):
    id = Column(db.String, primary_key=True)
    children = db.relationship("Child", backref="parent")

class Child(db.Model):
    id = Column(db.String, primary_key=True)
    parent_id = Column(db.ForeignKey("parent.id"), nullable=True)
```
Code
```
from database import db
from models import Parent, Child

import logging
logging.basicConfig()
logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)

parent = Parent(id="123")

db.session.add(parent)
db.session.flush()

child = Child(
    id="456",
    parent_id=parent.id
)

print(parent.children) # -> []

db.session.add(child)
db.session.flush()
print(parent.children) # -> [] if previous print is not there. If previous print is commented out, parent.children = [<Child...>]
```

## Error
The last print statement returns an empty array instead of the newly added `Child` object. I expect the second print statement to return the newly added child. 

Output from running above script:
```
[TID:140399841322816] INFO:root:Using override prod database URI
[TID:140399841322816] INFO:root:Using override events database URI
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:select version()
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:[raw sql] {}
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:select current_schema()
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:[raw sql] {}
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:show standard_conforming_strings
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:[raw sql] {}
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:INSERT INTO parent (id) VALUES (%(id)s)
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:[generated in 0.00022s] {'id': '123'}
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:SELECT child.id AS child_id, child.parent_id AS child_parent_id
FROM child
WHERE %(param_1)s = child.parent_id
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:[generated in 0.00020s] {'param_1': '123'}
[]
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:INSERT INTO child (id, parent_id) VALUES (%(id)s, %(parent_id)s)
[TID:140399841322816] INFO:sqlalchemy.engine.Engine:[generated in 0.00020s] {'id': '456', 'parent_id': '123'}
[]
```
## Versions
Flask-SQLAlchemy==2.5.1
SQLAlchemy==1.4.14
Python 3.8.10

## Additional context
If we comment out the first `print(parent.children)`, the second print works as expected. Here's the debug output for that:
```
[TID:140022678542144] INFO:root:Using override prod database URI
[TID:140022678542144] INFO:root:Using override events database URI
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:select version()
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:[raw sql] {}
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:select current_schema()
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:[raw sql] {}
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:show standard_conforming_strings
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:[raw sql] {}
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:INSERT INTO parent (id) VALUES (%(id)s)
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:[generated in 0.00022s] {'id': '123'}
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:INSERT INTO child (id, parent_id) VALUES (%(id)s, %(parent_id)s)
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:[generated in 0.00018s] {'id': '456', 'parent_id': '123'}
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:SELECT child.id AS child_id, child.parent_id AS child_parent_id
FROM child
WHERE %(param_1)s = child.parent_id
[TID:140022678542144] INFO:sqlalchemy.engine.Engine:[generated in 0.00021s] {'param_1': '123'}
[<Child 456>]
```

If this behavior is intended and we should be setting the relationship instead of the foreign key, are there thoughts on what kind of validation we could write to prevent this class of bugs?

**Comment by zzzeek** [comment, marked as answer]
this behavior is discussed in the FAQ at https://docs.sqlalchemy.org/en/14/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7 .

there are a variety of ways that people accommodate for this behavior.  the most common is that application code within the scope of a particular `Session` make use either exclusively with relationships, and not foreign key attributes, or the opposite, where code that deals with foreign key values directly does not try to access relationships within one operation, unless some form of expiration / refreshing is used to ensure the relationship-mapped attribute is loaded.   In your example above, this could be that you'd call `session.expire(parent, ['children'])` after that flush.    The usual way these bugs are avoided is by having unit tests for the code, since the behavior is typically deterministic given a particular code structure as opposed to being data driven and tests that cover the method will show if these issues are occurring.

Beyond that, there are several strategies that can be employed:

1. map all foreign key attributes using underscores
2. assign validation functions that prevent the value of a foreign key attribute from being changed outside of the space where the Session is in the flush process.   not exactly a public API but you'd look at the `_flushing` attribute on `Session`.    This is not something that people have asked for so it could be made more straightforward if there was demand for this feature.
3. similar to `#2` but more ambitious, assign events to foreign key attributes that then expire the relationships linked to that foreign key attribute, so that any access of the related attribute will immediately cause a reload (after an autoflush, if enabled).  For this, there's a recipe at https://github.com/sqlalchemy/sqlalchemy/wiki/ExpireRelationshipOnFKChange , but I don't get the impression this is very widely used and I also haven't tested it in awhile.   Adding this kind of feature is considered to be out of scope for SQLAlchemy itself.  `#2` above would be more feasible as a feature add.
